contributors/design-proposals/scheduling/scheduling-policy.md

```diff
@@ -0,0 +1,516 @@
+
+_Authors: @arnaudmz, @yastij_
+
+_Reviewers: @bsalamat, @liggitt_
+
+# Objectives
+
+-  Define the concept of scheduling policies
+-  Propose their initial design and scope
+
+## Non-Goals
+
+-  How taints / tolerations work.
+-  How NodeSelector works.
+-  How node / pod affinity / anti-affinity rules work.
+-  How several schedulers can be used within a single cluster.
+-  How priority classes work.
+-  How to set defaults in Kubernetes.
+
+# Background
+
+During real-life Kubernetes architecturing we encountered contexts where role-isolation (between administration and simple namespace usage in a multi-tenant context) could be improved. So far, no restriction is possible on toleration, priority class usage, nodeSelector, anti-affinity depending on user permissions (RBAC).
+
+Identified use-cases aim to ensure that administrators have a way to restrict users or namespaces, it allows administrators to:
+
+-  Restrict execution for specific applications (which are namespace scoped) into certain nodes
+-  Create policies that prevent users from even attempting to schedule workloads onto masters to maximize security
+-  require that a pods under a namespace run on dedicated nodes
+-  Restrict usage of some `PriorityClass`
+-  Restrict usage to a specific set of schedulers.
+-  placing pods on specific nodes (master roles for instance).
+-  using specific priority classes.
+-  expressing pod affinity or anti-affinity rules.
+
+Also Mandatory (and optionally default) values must also be enforced by scheduling policies in case of:
+
+-  mutli-arch (amd64, arm64) of multi-os (Linux, Windows) clusters
+-  multi-az / region / failure domain clusters
+
+# Overview
+
+### syntaxic standards
+
+```yaml
+apiVersion: policy/v1alpha1
+kind: <POLICY_KIND>
+metadata:
+  name: <POLICY_NAME>
+spec:
+  bindingMode: <BINDING_MODE>            # Describes a bindingMode (any or all)
+  namespaces:                            # a list of namespaces that the policy targets (optional)
+    - ns1
+    - ns1
+  namespaceSelector:                     # a selector to match a set of namespaces (optional)
+    key: value
+  rules:                                 # rules that must be satisfied (optional)
+    fieldA:                              # name of the field (optional)
+      - match:                             # Describes how the rule is matched (required)
+        - elt1                           # elements here could be objects like tolerations or strings like SchedulerName
+        action:                            # Describes the action that should be taken (allowed, denied or required)
+
+```
+
+### policy composition
+
+`bindingMode` Describes how policies should be composed:
+
+- any: any policy that has its rules satisfied
+- all: all policies MUST have its rules satisfied
+
+if we have a heterogenous bindingMode across policies (i.e. some policies with any and others with all).
+Then the most restrictive one (i.e the all bindingMode) is applied.
```

Comment:
I think we should go with the [firewall analogy](https://github.com/kubernetes/community/pull/1950#discussion_r176531090) instead. It's both more familiar and more powerful:

- Default behavior is to allow everything (I.e. if no policy is matched)
- Policy has an `Action` that is "Allow" (or accept?), "Deny" (or reject?), (and maybe "Default")
- Policy has a priority.
- The action is taken on the first policy (priority order) that matches (short circuit): i.e. if it's allow, the pod is accepted. If it's deny, the pod is rejected. Deny takes priority over allow when there's a conflict.

The "any" and "all" modes map to this:
- allow+any: higher priority allow policies, with a low priority default-deny
- allow+all: higher priority denies (it may be useful to have an "invert" option that applies to everything that _doesn't_ match)
- deny+any: higher priority deny policies
- deny+all: higher priority allow (inverse), default deny

Reply:
I see, this isn’t much different from what we have now. But basically, using this approach we would have action at a top level and not for rule IIUC ?

Reply:
I like the linear approach with the firewall model since it's easy to reason through the computation. What I don't like is that I worry it would encourage users to stack on more and more policies in order to get what they want rather than think about the policies more comprehensively.

OTOH reading through some of the formal security models developed, I don't know that we can have a set of policies as such without a partial ordering.

I've been trying to reason through a few ways to define policy and I struggle because I don't know where kubernetes should lie abstractly. Should we define policy so as to have as few layers as possible, even if it requires tooling to make that useful for admins? Or should it be as flexible as possible for defining within the native abstractions?

Reply:
Here's what I've been thinking about on policy composition and conflicts:

#### Policy Conflicts

Conflicts arise when two policies prescribe different results for the same subject. Conflicts often arise when policies can be defined for any grouping or subset of possible and existing objects.

#### Resolution Options
1. Policies are defined on a topology, like a formal one
    - Example: Role bindings in RBAC
2. Define a priority ordering
   1. On policies
       1. Explicitly, user specified. e.g. firewalls, Priority Classes
       2. Arbitrarily, though hopefully consistently
               - Alphanumeric sort
               - By timestamp
               - By permissiveness, e.g. - NetworkPolicy (“fail-closed”, use most restrictive policy), Authorization hooks (“fail-open”)
   2. On classifiers
        1.  e.g. Security clearance levels
        2. hierarchical labels
3. Use a combination
     - For example, PriorityClasses
       - A user doesn’t need to know about the relative priority of classes, just the category to put the pod in
       - An admin can manage priorities independent of the particular pods.
---

I lean towards option 3 like PriorityClasses, maybe something kinda like this https://gist.github.com/ericavonb/8ce70d9c9e300b0eec8842eb44a4cd49

@tallclair @yastij does any of that make sense? thoughts?

Reply:
1. I don't understand this. How does it resolve conflicts? RBAC doesn't have conflicts (pure whitelist). Conflicts only arise when you have mixed action types.
2.
    i. User specified priority is insufficient without additional rules to ensure that a conflict cannot arise on a single priority (e.g. only one policy per priority, or strict ordering between actions (deny takes precedence over allow))
    ii. I don't understand how this resolves conflicts. Or is the idea that there is only a single policy that maps to an object? You still need to resolve conflicts in the mapping?
3. Is this like (2.i), but rather than specifying arbitrary numerical priorities, there are priority classes? I was thinking about something like that the other day, e.g.
        - "exceptions" (highest) - e.g. no restrictions on kube-system & other privileged namespaces
        - "cluster priority" - specify general cluster-wide policy
        - "application priority" - additional namespace or application specific constraints
        - "default priority" (lowest) - set defaults & other things that can still be overridden at application priority.
    On the one hand, this simplifies management by being more perscriptive and prevents users from getting into a runaway scenario (layering on too many policies until it becomes unmanageable). On the other hand, the user needs to memorize the ordering, as opposed to implicitly ordered numeric values.

Reply:
@tallclair - yes we can have a priory based approach as discussed and then use an action per policy with multiple matchers.